### PR TITLE
Fixes #13345 : UI Automation: Write an automated test using Cypress for "Selected options with long usernames are not cut off in the RHS"

### PR DESCRIPTION
**Pull request by M-ZubairAhmed**
#### Summary
Test to check if long usernames are truncated in options in rhs

#### Ticket Link
Fixes : https://github.com/mattermost/mattermost-server/issues/13345
Jira : https://mattermost.atlassian.net/browse/MM-21038

### CODE REVIEW

**Comment by M-ZubairAhmed** [comment]
This PR is blocked until issue addressed raised here https://github.com/mattermost/mattermost-server/issues/13345#issuecomment-569225732 are addressed

**Comment by saturninoabril** [comment]
@M-ZubairAhmed The long username is permissible on server but has limitation on client only.  I'd suggest to continue with the original test requirement with my comment on how to create new user with long username.

File: e2e/cypress/integration/interactive_menu/basic_options_spec.js [hunk 1]
```diff
@@ -60,6 +60,19 @@ describe('Interactive Menu', () => {
...
+
+        // Lets check if user already exist
+        cy.apiGetUsers([userWithLongName.username]).then((usersResponse) => {
+            const {body} = usersResponse;
+
+            // If we dont have 'long user' in the app, we create user now
+            if (body.length === 0) {
+                cy.createNewUser(userWithLongName);
+            }
+        });
```
[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply e2e/cypress/integration/interactive_menu/basic_options_spec.js]: But as per specs shouldnt i be creating a long username of 64 chars?

[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply e2e/cypress/integration/interactive_menu/basic_options_spec.js]: as can be seen here https://github.com/mattermost/mattermost-server/issues/13345#issuecomment-568731249 , 22 chars doesnt justify what the test require about not truncating

[the remark this answers is not in the document]
Comment (saturninoabril) [reply e2e/cypress/integration/interactive_menu/basic_options_spec.js]: Yup, you should be able to create username with 64 chars.  I did that with,
```javascript
cy.getCurrentTeamId().then((teamId) => {
    const username = `aaa_64chars_user_abcdefghijklmnopqrstuvqxyz01234567${Date.now()}`;

    cy.createNewUser({username}, [teamId]).as('longUser');
});
```

Screen Shot 2020-01-02 at 4 52 02 PM

[the remark this answers is not in the document]
Comment (M-ZubairAhmed) [reply e2e/cypress/integration/interactive_menu/basic_options_spec.js]: how was it that previously i was not able to create and now i am able to make user with 64 chars ? was the bug fixed

**Comment by M-ZubairAhmed** [comment]
This PR is blocked until team discussion for how does long username should look like

**Comment by saturninoabril** [comment]
https://mattermost.atlassian.net/browse/MM-21467 - ticket filed on expected behavior of truncated and not cut-off.

**Comment by M-ZubairAhmed** [comment]
Closing this until the issue above is solved

**Comment by saturninoabril** [comment]
@M-ZubairAhmed fyi, selected long username is now truncated.  Fixed by https://github.com/mattermost/mattermost-webapp/pull/4671

**Comment by M-ZubairAhmed** [comment]
@josephbaylon Hi, thank you for reviewing my PR, but do you mind checking if webhook server is open ?

**Comment by josephbaylon** [comment]
@M-ZubairAhmed I'm not sure what you're referring to. Are you referring to the service settings in Mattermost server? EnableIncomingWebhooks/EnableOutgoingWebhooks (set to true) ?

**Comment by M-ZubairAhmed** [comment]
oh i am taking about ` Note: This test requires webhook server running. Initiate `npm run start:webhook` to start.`

**Comment by josephbaylon** [comment]
@M-ZubairAhmed When I have the webhook turned on I get this different failure consistently
Screen Shot 2020-01-27 at 10 23 50 PM

**Comment by josephbaylon** [comment]
@saturninoabril Are you able to run this test fine? If so, I'll defer to your results

**Comment by M-ZubairAhmed** [comment]
@josephbaylon do you mind sharing the complete screen shot ?

**Comment by josephbaylon** [comment]
@M-ZubairAhmed Here's what I've observed. If I run it by itself, the test passes but if I run it with the other tests in the file, it fails. For some reason, when run with other tests, the webhook post for user options doesn't get posted. I tried adding a timeout but still the same. Maybe you could run the whole test and see if you get the same behavior.

Screen Shot 2020-01-27 at 11 55 04 PM

**Comment by M-ZubairAhmed** [comment]
@josephbaylon oh yes i can see that now, it looks like there is some issue with the webhook. Do you and @saturninoabril know how this can be solved ?

**Comment by saturninoabril** [comment]
For some reason, all tests passed on me the first time I reviewed this and now seeing the same failing test as yours.  I'll take a look.

**Comment by M-ZubairAhmed** [comment]
thank you @saturninoabril

**Comment by saturninoabril** [comment]
I'm seeing invalid session log on server. It may be due to `cy.createNewUser` that sign out then sign in `sysadmin`.  For long term fix, we might need to revisit `cy.createNewUser`.  Short fix for this PR is to move `cy.createNewUser` at before hook.
```javascript
...
let longUsername;

describe('Interactive Menu', () => {
    before(() => {
        // Set required ServiceSettings
        const newSettings = {
            ServiceSettings: {
                AllowedUntrustedInternalConnections: 'localhost',
                EnablePostUsernameOverride: true,
                EnablePostIconOverride: true,
            },
        };
        cy.apiUpdateConfig(newSettings);

        cy.apiLogin('sysadmin').visit('/ad-1');
        longUsername = `name-of-64-abcdefghijklmnopqrstuvwxyz-123456789-${Date.now()}`;
        cy.getCurrentTeamId().then((teamId) => {
            // # Create a new user with 64 chars length
            cy.createNewUser({username: longUsername}, [teamId]);
        });

        // # Login as sysadmin and ensure that teammate name display setting is set to default 'username'
        cy.apiLogin('sysadmin');
        ...
```

**Comment by M-ZubairAhmed** [comment]
thank you @saturninoabril for swiftly identifying the bug, I will send the update.
In future please refer the issue for the fix to me, i will be happy to help

**Comment by M-ZubairAhmed** [comment]
Hi, i fixed the test as per Saturnino's comment

File: package-lock.json [hunk 2]
```diff
@@ -11938,6 +11941,12 @@
         "postcss": "^7.0.14"
       }
     },
+    "icu4c-data": {
+      "version": "0.64.2",
+      "resolved": "https://registry.npmjs.org/icu4c-data/-/icu4c-data-0.64.2.tgz",
+      "integrity": "sha512-BPuTfkRTkplmK1pNrqgyOLJ0qB2UcQ12EotVLwiWh4ErtZR1tEYoRZk/LBLmlDfK5v574/lQYLB4jT9vApBiBQ==",
+      "dev": true
+    },
```
Comment (saturninoabril) [inline package-lock.json]: Would be nice if the changes are reverted since it's not related to the PR itself.

**Comment by josephbaylon** [comment]
@saturninoabril I'm not sure what's happening with the untagged-build failure.. says `upload-s3 - Unauthorized` Do you know how to resolve this?

**Comment by saturninoabril** [comment]
If no outstanding issue on CI, we may try updating the branch and see if it passed.  Else, we may ask help from Build team.

**Pull request merged**